x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -302,6 +273,51 @@ public void authenticate(final AuthenticationToken authenticationToken, final Ac
         }
     }
 
+    private Void doAuthenticate(ActionListener<AuthenticationResult<User>> listener, JwtAuthenticationToken jwtAuthenticationToken) {
+        final String tokenPrincipal = jwtAuthenticationToken.principal();
+
+        // Authenticate client: If client authc off, fall through. Otherwise, only fall through if secret matched.
+        final SecureString clientSecret = jwtAuthenticationToken.getClientAuthenticationSharedSecret();
+        try {
+            JwtUtil.validateClientAuthentication(clientAuthenticationType, clientAuthenticationSharedSecret, clientSecret);
+            logger.trace("Realm [{}] client authentication succeeded for token=[{}].", name(), tokenPrincipal);
+        } catch (Exception e) {
+            final String msg = "Realm [" + name() + "] client authentication failed for token=[" + tokenPrincipal + "].";
+            logger.debug(msg, e);
+            listener.onResponse(AuthenticationResult.unsuccessful(msg, e));
+            return null;
+        }
+
+        final BytesArray jwtCacheKey = isCacheEnabled() ? new BytesArray(jwtAuthenticationToken.getUserCredentialsHash()) : null;
+        if (jwtCacheKey != null) {
+            final User cachedUser = tryAuthenticateWithCache(tokenPrincipal, jwtCacheKey);
+            if (cachedUser != null) {
+                if (delegatedAuthorizationSupport.hasDelegation()) {
+                    delegatedAuthorizationSupport.resolve(cachedUser.principal(), listener);
+                } else {
+                    listener.onResponse(AuthenticationResult.success(cachedUser));
+                }
+                return null;
+            }
+        }
+
+        // Validate JWT: Extract JWT and claims set, and validate JWT.
+        jwtAuthenticator.authenticate(jwtAuthenticationToken, ActionListener.wrap(claimsSet -> {
+            // We need another doPrivileged call here, since ActionListener.wrap does not preserve the outer context
+            SpecialPermission.check();
+            AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
```

Comment:
This is the only real diff here, everything else is just code moving.

Reply:
Mostly for my education: why does `ActionListener.wrap` lose the outer context, i.e. why is `doPrivileged` needed again? Does it mean we need add `doPrivileged` each time we use `ActionListener.wrap`? This feels quite trappy. If any downstream code is refactored to use `wrap`, it is hard to expect the person who makes the change to know about the need for doPrivileged?

Reply:
My high level understanding is that security is thread context dependent (@ChrisHegarty would you be able to elaborate?) 

>  Does it mean we need add doPrivileged each time we use ActionListener.wrap? This feels quite trappy. 

Yes unfortunately. I don't really see a way around this though. I'm still not positive if this is a complete blocker here or not. Integration tests will catch this, but unit tests will not (see the second part of my comment [here](https://github.com/elastic/elasticsearch/pull/96483#discussion_r1230541520)). Let's discuss on Slack.

Reply:
The security manager is essentially thread-sensitive, which is fundamentally at odds with async programming - hence the need for the above. This is not new, and has been the situation forever. Trappy, maybe, but any Java dev using the security manager (<1%) knows this. There are ways to capture the calling context and assert it during the callback, but I see no point doing this here - it just adds more unnecessary complication.

The security manager is not a perfect API, or even a good concept for programs written in the last 10+ years, hence its deprecation.

Reply:
Just adding a note here for posterity:

`ActionListener.wrap` causes permission issues when its `onResponse`  method (or `onFailure` for that matter) is called from an unprivileged context. To illustrate, suppose we have (apologies for mix of Java and pseudo-code):

```
// we have a security policy for the pack security plugin granting `ExamplePermission`

// all calls below made from org.elasticsearch.xpack.security code

var l = ActionListener.wrap(
    ignored -> AccessController.checkPermission(new ExamplePermission()),
    ex -> ...
);

...

// this works just fine
AccessController.doPriviliged(() -> {
    ...
    l.onResponse(null)
  }
);

...

// this fails with an access denied exception, since we are outside of a doPrivileged block
l.onResponse(null)
```